rdmd.d

```diff
@@ -916,3 +891,52 @@ void yap(size_t line = __LINE__, T...)(auto ref T stuff)
     debug stderr.writeln(line, ": ", stuff);
     else stderr.writeln(stuff);
 }
+
+/**
+Used to wrap filesystem operations that should also be logged with the
+yap function. These functions allow the the filenane to be given once
+so the log statement always matches the operation. Using it also guarantees
+you won't forget to include a `yap` alongside any file operation you want
+to be logged.
+*/
+static struct Yap
+{
+    static import std.file;
+
+    static auto opDispatch(string func, T...)(T args)
+    {
+        static if (func.endsWith("IfLive"))
+        {
+            enum fileFunc = func[0 .. $ - "IfLive".length];
+            enum skipOnDryRun = true;
+        }
+        else
+        {
+            enum fileFunc = func;
+            enum skipOnDryRun = false;
+        }
+
+        static if ([fileFunc].findAmong(["exists", "timeLastModified", "isFile"]))
+            yap("stat ", args[0]);
+        else static if (fileFunc == "rename")
+            yap("mv ", args[0], " ", args[1]);
+        else static if ([fileFunc].findAmong(["remove", "mkdirRecurse", "rmdirRecurse", "dirEntries", "write"]))
+            yap(fileFunc, " ", args[0]);
+        else static assert(0, "Yap.opDispath has not implemented " ~ fileFunc);
+
+        static if (skipOnDryRun)
+        {
+            if (dryRun)
+                return;
+        }
+        mixin("return std.file." ~ fileFunc ~ "(args);");
+    }
+
+    // Special case, combine these 2 operations so you don't end up with
+    // multiple yap statements in the log for the same file.
+    static bool existsAsFile(const(char)[] name)
+    {
+        yap("stat ", name);
+        return std.file.exists(name) && std.file.isFile(name);
+    }
```

Comment:
OK, I understand what you mean. But it's not obvious because the layers (logging, dispatching, and implementation) are conflated here.

How about this:
```d
struct RealFilesystem
{
static:
	import std_file = std.file;
	alias std_file this;

	/// Create blank file.
	void touch(string fileName)
	{
		std_file.write(fileName, "");
	}

	/// Create blank file.
	bool existsAsFile(string name)
	{
		 return std_file.exists(name) && std_file.isFile(name);
	}
}
```

Then in `opDispatch`, forward to `RealFilesystem` instead of `std.file`.

Reply:
done, I chose `Filesystem` instead of `RealFilesystem` though

Reply:
Err... where is it? :)

Reply:
Sorry, I thought you were saying...rename `Yap` to `RealFilesystem`....

Reply:
That too! :D

Reply:
now done :)

Reply:
`DirectFilesystem` was a pretty clever idea...nice